File: old_contents/boj_codes/Python3/1260.py
#https://www.acmicpc.net/problem/1260
# 20200911

import sys

N_node, N_edge, init_node = map(int, sys.stdin.readline().split())

class Graph:
    def __init__(self, N_node):
        self.N_node = N_node
        self.nodes = []
        for _ in range(self.N_node + 1):
            self.nodes.append([0] * (self.N_node + 1))

    def add_edge(self, node1, node2):
        self.nodes[node1][node2] = 1
        self.nodes[node2][node1] = 1

    def is_connected(self, node1, node2):
        return True if self.nodes[node1][node2] == 1 else False

    def init_search(self):
        self.visited = [0] * (self.N_node + 1)

    def BFS(self, init_node):
        queue = [init_node]
        self.visited[init_node] = 1

        while len(queue) != 0:
            node = queue[0]
            queue.pop(0)

            print(node, end=' ')

            for i in range(1, self.N_node + 1):
                if self.nodes[node][i] == 1:
                    if self.visited[i] == 0:
                        self.visited[i] = 1
                        queue.append(i)

    def DFS(self, node):
        print(node, end=' ')
        self.visited[node] = 1

        for i in range(1, self.N_node + 1):
            if self.nodes[node][i] == 1:
                if self.visited[i] == 0:
                    self.DFS(i)



graph = Graph(N_node)
for _ in range(N_edge):
    n1, n2 = map(int, sys.stdin.readline().split())
    graph.add_edge(n1, n2)

graph.init_search()
graph.DFS(init_node)
print()
graph.init_search()
graph.BFS(init_node)

File: old_contents/boj_codes/C++17/1260.cpp
// https://www.acmicpc.net/problem/1260
// First Written : 20201229
// Last Modified : 20201229

#include <cstdio>
#include <deque>

using namespace std;

struct Edge{
    int start, end;
};

class Graph{
private:
    bool** adj_list;
    deque<int> queue;
    bool* visited;
    int N_node;
    int cur_node;

public:
    Graph(int N_node){
        adj_list = new bool*[N_node + 1];
        for(int i = 1; i <= N_node; i++){
            adj_list[i] = new bool[N_node + 1]();
        }
        visited = new bool[N_node + 1]();
        this->N_node = N_node;
    }

    ~Graph(){
        for(int i = 1; i <= N_node; i++){
            delete [] adj_list[i];
        }
        delete [] adj_list;
        delete [] visited;
    }

    void init_visited(){
        delete [] visited;
        visited = new bool[this->N_node + 1]();
    }

    void add_edge(const Edge e){
        adj_list[e.start][e.end] = true;
        adj_list[e.end][e.start] = true;
    }


    void BFS(int init_node){
        queue.push_back(init_node);
        visited[init_node] = true;

        while(queue.size() != 0){
            cur_node = queue[0];
            queue.pop_front();

            printf("%d ", cur_node);

            for(int node = 1; node < this->N_node + 1; node++){
                if((!visited[node]) && adj_list[cur_node][node]){
                    visited[node] = true;
                    queue.push_back(node);
                }
            }
        }
    }

    void DFS(int cur_node){
        printf("%d ", cur_node);
        visited[cur_node] = true;

        for(int node = 1; node < this->N_node + 1; node++){
            if((!visited[node]) && adj_list[cur_node][node]){
                DFS(node);
            }
        }
    }

};

int main(){
    int N_node, N_edge, start_node;
    scanf("%d %d %d", &N_node, &N_edge, &start_node);

    Graph graph = Graph(N_node);
    Edge e;
    while(N_edge--){
        scanf("%d %d", &e.start, &e.end);
        graph.add_edge(e);
    }

    graph.DFS(start_node);
    printf("\n");
    graph.init_visited();
    graph.BFS(start_node);

    return 0;
}